src/java.base/unix/native/libnio/fs/UnixCopyFile.c

```diff
@@ -47,15 +51,36 @@ static void throwUnixException(JNIEnv* env, int errnum) {
 }
 
 /**
- * Transfer all bytes from src to dst via user-space buffers
+ * Transfer all bytes from src to dst within the kernel if possible (Linux),
+ * otherwise via user-space buffers
  */
 JNIEXPORT void JNICALL
 Java_sun_nio_fs_UnixCopyFile_transfer
     (JNIEnv* env, jclass this, jint dst, jint src, jlong cancelAddress)
 {
-    char buf[8192];
     volatile jint* cancel = (jint*)jlong_to_ptr(cancelAddress);
 
+#if defined(__linux__)
+    // Attempt to transfer within the kernel
+    ssize_t bytes_sent;
+    do {
+        // sendfile() will transfer at most 0x7ffff000 bytes
+        RESTARTABLE(sendfile64(dst, src, NULL, 0x7ffff000), bytes_sent);
+        if (cancel != NULL && *cancel != 0) {
+            throwUnixException(env, ECANCELED);
+            return;
+        }
+    } while (bytes_sent > 0);
+
+    // sendfile() returns zero at src EOF
+    if (bytes_sent == 0) {
+        return;
+    }
+#endif
```

Comment:
This version looks much better but it missing the error checking. If bytes_sent is -1 then we should throw UnixException with the errno like the existing code.

What would you think about setting count to 1MB or 4MB so that there is at least some chance for cancellation when the file is huge.

Given that the new implementation will call sendfile64 until EOF then there is no reason to fall through to the old implementation, the old implementation goes into the #else ...

Reply:
Changing `count` to 1MB and 4MB instead of 2147479552 bytes somewhat surprisingly did not have a huge effect on the throughput for the two largest transfer sizes tested.